core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,453 @@
+    for (SocketAddress oldAddr : oldAddrs) {
+      if (!newAddrs.contains(oldAddr)) {
+        subchannels.get(oldAddr).shutdown();
+        states.remove(subchannels.remove(oldAddr));
+      }
+    }
+
+    if (oldAddrs.size() == 0 || currentState == CONNECTING || currentState == READY) {
+      // start connection attempt at first address
+      updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+      requestConnection();
+
+    } else if (currentState == IDLE) {
+      // start connection attempt at first address when requested
+      SubchannelPicker picker = new RequestConnectionPicker(this);
+      updateBalancingState(IDLE, picker);
+
+    } else if (currentState == TRANSIENT_FAILURE) {
+      // start connection attempt at first address
+      requestConnection();
+    }
+
+    return true;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    for (Subchannel subchannel : subchannels.values()) {
+      subchannel.shutdown();
+    }
+    subchannels.clear();
+    // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+    // for time being.
+    updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+  }
+
+  void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    ConnectivityState newState = stateInfo.getState();
+    // Shutdown channels/previously relevant subchannels can still callback with state updates.
+    // To prevent pickers from returning these obselete subchannels, this logic
+    // is included to check if the current list of active subchannels includes this subchannel.
+    if (!subchannel.equals(subchannels.get(subchannel.getAddresses().getAddresses().get(0)))) {
+      return;
+    }
+    if (newState == SHUTDOWN) {
+      return;
+    }
+    if (newState == IDLE) {
+      helper.refreshNameResolution();
+    }
+    // If we are transitioning from a TRANSIENT_FAILURE to CONNECTING or IDLE we ignore this state
+    // transition and still keep the LB in TRANSIENT_FAILURE state. This is referred to as "sticky
+    // transient failure". Only a subchannel state change to READY will get the LB out of
+    // TRANSIENT_FAILURE. If the state is IDLE we additionally request a new connection so that we
+    // keep retrying for a connection.
+
+    // With the new pick first implementation, individual subchannels will have their own backoff
+    // on a per-address basis. Thus, iterative requests for connections will not be requested
+    // once the first pass through is complete.
+    // However, every time there is an address update, we will perform a pass through for the new
+    // addresses in the updated list.
+    if (currentState == TRANSIENT_FAILURE) {
+      if (newState == CONNECTING) {
+        // each subchannel is responsible for its own backoff
+        return;
+      } else if (newState == IDLE) {
+        requestConnection();
```

Comment:
this is sticky transient failure, so we should return immediately.

Reply:
I think we discussed that we never go from TRANSIENT_FAILURE to IDLE since when coming out of TF, a subchannel goes to CONNECTING. 

Would we want to remove this altogether? I think you previously told me to include the requestConnection() logic in case a state is skipped?

Reply:
That is still true, immediately after requestConnection, return. Otherwise helper is notified with state changes.